torch/optim/_functional.py

```diff
@@ -322,3 +322,47 @@ def adamax(params: List[Tensor],
         clr = lr / bias_correction
 
         param.addcdiv_(exp_avg, exp_inf, value=-clr)
+
+
+def asgd(params: List[Tensor],
+         grads: List[Tensor],
+         axs: List[Tensor],
+         mus: List[float],
+         etas: List[float],
+         state_steps: List[int],
+         *,
+         lr: float,
+         weight_decay: float,
+         alpha: float,
+         t0: float,
+         lambd: float):
+    r"""Functional API that performs asgd algorithm computation.
+
+    See :class:`~torch.optim.ASGD` for details.
+    """
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        mu = mus[i]
+        ax = axs[i]
+        eta = etas[i]
+        step = state_steps[i]
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # decay term
+        param.mul_(1 - lambd * eta)
+
+        # update parameter
+        param.add_(grad, alpha=-eta)
+
+        # averaging
+        if mu != 1:
+            ax.add_(param.sub(ax).mul(mu))
+        else:
+            ax.copy_(param)
+
+        # update eta and mu
+        eta = (lr / math.pow((1 + lambd * lr * step), alpha))
+        mu = 1 / max(1, step - t0)
```

Comment:
so the CI failure looks real, I think it's because `eta` and `mu` are not updated in the optimizer's state, since they are just floats, not tensors, when passing into this API, it's merely a copying rather than using reference semantics like tensor. We should think of how to handle the non tensor optimizer states like float or int

Reply:
Thank you for pointing this out. Indeed  its not a great idea to give List[float] to the functional, as we decided in our offline discussion, i modified the code, so we could do the update inside asgd.py rather than functional